[WIP] Extend Wikidata extension to support arbitrary Wikibase instances

afkbrb opened the pull request:
see #1640

wetneb commented:
(@afkbrb by the way, feel free to divert from the roadmap I outlined in #1640 as much as you feel appropriate - I trust that you will come up with very sensible design choices.)

afkbrb commented:
@wetneb @tfmorris I still need to spend a few days studying about Wikibase and also other things related before coding, I'll reach you here if I need your help.

I plan to wirte a prototype for https://test.wikidata.org/ first.

afkbrb commented:
> I plan to wirte a prototype for https://test.wikidata.org/ first.

test.wikidata doesn't seem to have a SPARQL query service, so I cannot run the reconciliation service for it. I think I should set up a local wikibase and run related services for it, so I can test on it.

afkbrb commented:
@wetneb @tfmorris This is my design for the workflow of the Wikibase support below, I'm going to reuse lots of code of adding/deleting/listing/selecting/saving/restoring reconciliation services.

1. Load saved Wikibase instances from preferences when the schema page is opened for the first time.

2. Add a new Wikibase instance (optional). Just like adding a new reconciliation service, the frontend tries fetching the manifest with the URL first. If successful, sends the manifest to the backend. The backend saves the manifest to the preferences. The service added is selected automatically and we go to the next step.

3. Select the target Wikibase instance. Both the frontend and the backend should be updated according to the Wikibase. The frontend should inform the user that the target Wikibase is changed and that the user should reconcile the data to the new Wikibase first. The backend updates the ManifestManager. The ManifestManager should have some getters to get necessary information (such as the MediaWiki API endpoint, Qid and Pid of the constraint types) of the current Wikibase. The backend is updated before editing the schema, so the scrutinizers can work properly.

4. Edit the schema.

5. Upload edits. If the user is not logged in to the target Wikibase, show the login dialog. The dialog UI should also be changed. Mentions to Wikidata should be changed to the target Wikibase. The Wikidata logo should also be changed to the Wikibase logo. I think the "logging in with owner-only consumer" option introduced by #2661 should be hidden if the target Wikibase is not Wikidata, since not all Wikibase instances support that, and also the [wiki](https://github.com/OpenRefine/OpenRefine/wiki/Wikidata-owner-only-consumer) we wrote is designated to Wikidata.

I'll figure out more details later. Is there any potential problem with this design you can see for now?

BTW, the first evaluation of GSoC has started :)

wetneb commented:
This looks like a great plan!

tfmorris commented:
@afkbrb That sounds like a good plan. Don't worry, we won't forget your evaluation! 

@wetneb Do we know of any friendly Wikibase admins who would be potential users that we could request feedback from?

wetneb commented:
@tfmorris yes! https://daap.bannerrepeater.org/wiki/Main_Page is a good example, run by Lozana Rossenova who is on our steering committee. They are in the process of setting up a reconciliation endpoint for it, which is the first step. I can do intros when needed.

afkbrb commented:
Hi @wetneb , in [WikidataConstraintFetcher](https://github.com/OpenRefine/OpenRefine/blob/master/extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java) , the constraint types of Wikidata are hard-coded. Can we fetch them via the MediaWiki API or other web services? If we can't, we need to add all of them manually to the manifest of a Wikibase, so the manifest can be lengthy and complex.

wetneb commented:
Correct, I do not think the configuration is exposed publicly at the moment. For now the only solution I can see is to include all the Qids and Pids in the manifest, sadly. These could be optional (since Wikibase-Quality-Constraints might not be installed on the target instance, or they might not use all constraint types).

afkbrb commented:
@wetneb @tfmorris I think we should migrate some methods from [WikidataConstraintFetcher](https://github.com/OpenRefine/OpenRefine/blob/master/extensions/wikidata/src/org/openrefine/wikidata/qa/WikidataConstraintFetcher.java) to corresponding scrutinizers. 

In that calss, many methods are designated to a specific scrutinizer. For example, the `getFormatRegex` method is used only in `FormatScrutinizer`, the `getInversePid` method is only used in `InverseConstraintScrutinizer`, etc. These methods should be private/protected  in scrutinizers instead of being exposed in public in the fetcher. Methods such as `getSingleConstraint` and `getConstraintStatements` should be kept though, so we can still use the cache functionality of `WikidataConstraintFetcher`. 

The migration is helpful for the arbitrary Wikibase support. The idea is that many scrutinizers depend on some constraints information (for example, the `FormatScrutinizer` depends on `FORMAT_CONSTRAINT_QID`(Q21502404 for Wikidata) and `FORMAT_REGEX_PID`(P1793 for Wikidata)), and after the migration, we can easily know what the dependencies of a scrutinizer are (since they are restricted in the scrutinizer then) and check if they are missing in the manifest. If the dependencies are missing, we can simply skip the scrutinizer. For example, we can adjust the code of [FormatScrutinizer#scrutinize](https://github.com/OpenRefine/OpenRefine/blob/master/extensions/wikidata/src/org/openrefine/wikidata/qa/scrutinizers/FormatScrutinizer.java#L103):

```java
    public void scrutinize(Snak snak, EntityIdValue entityId, boolean added) {
        if (manifest.getFormatConstraintQid() == null || manifest.getFormatRegexPid() == null) return; // skip if any dependency is missing
        
        if (snak.getValue() instanceof StringValue) {
         ... // original code
        }
    }
```

afkbrb commented:
The commit above shows what I mean. I haven't added dependencies check of scrutinizers yet.

wetneb commented:
@afkbrb yes indeed, that is the migration @darecoder1999 has been working on for the past few weeks. She will delete those methods from the `ConstraintFetcher` interface soon.

afkbrb commented:
Great! @darecoder1999 's work will make it easier to merge the code here.

tfmorris commented:
Those methods are removed in c5e6ac9f9016410dc6b44acc7a700490c4f2eca1. If you rebase your branch against the current HEAD of `master`, you should pick up the latest.

afkbrb commented:
Thanks for reminding!

afkbrb commented:
@wetneb Is [schema-alignment-dialog.html](https://github.com/OpenRefine/OpenRefine/blob/master/extensions/wikidata/module/scripts/dialogs/schema-alignment-dialog.html) needed? I couldn't find any usage of it.

wetneb commented:
@afkbrb indeed it seems to be a leftover of the previous UI where the schema was in its own dialog rather than in a tab.

afkbrb commented:
I've deleted it and the schema tab works fine. I also renamed [schema-alignment-dialog.js](https://github.com/OpenRefine/OpenRefine/blob/master/extensions/wikidata/module/scripts/dialogs/schema-alignment-dialog.js) to `schema-alignment.js`, renamed `SchemaAlignmentDialog` to `SchemaAlignment` and moved that file out of the `dialogs` folder.

tfmorris commented on extensions/wikidata/src/org/openrefine/wikidata/qa/scrutinizers/RestrictedValuesScrutinizer.java:
```diff
@@ -15,19 +17,32 @@
+    public boolean prepareDependencies() {
+        OneOfConstraint oneOfConstraint = constraints.getOneOfConstraint();
+        NoneOfConstraint noneOfConstraint = constraints.getNoneOfConstraint();
+        if (oneOfConstraint == null) return false;
+        if (noneOfConstraint == null) return false;
```
```suggestion
        if (oneOfConstraint == null || noneOfConstraint == null) {
            return false;
        }
```

tfmorris commented on extensions/wikidata/src/org/openrefine/wikidata/qa/scrutinizers/SingleValueScrutinizer.java:
```diff
@@ -45,17 +47,28 @@
+        SingleValueConstraint singleValueConstraint = constraints.getSingleValueConstraint();
+        SingleBestValueConstraint singleBestValueConstraint = constraints.getSingleBestValueConstraint();
+        if (singleValueConstraint == null) return false;
+        if (singleBestValueConstraint == null) return false;
```
combine as above

tfmorris commented on extensions/wikidata/src/org/openrefine/wikidata/qa/scrutinizers/SingleValueScrutinizer.java:
```diff
@@ -45,17 +47,28 @@
+        if (singleBestValueConstraint == null) return false;
+        singleValueConstraintQid = singleValueConstraint.getQid();
+        singleBestValueConstraintQid = singleBestValueConstraint.getQid();
+        return false;
```
Doesn't this mean that this will never get used?

afkbrb replied:
This is obviously a mistake :)

tfmorris commented on extensions/wikidata/src/org/openrefine/wikidata/schema/WikibaseSchema.java:
```diff
@@ -57,38 +57,39 @@
      * @return the site IRI of the Wikibase instance referenced by this schema
      */
-    @JsonIgnore
+    @JsonProperty("entityPrefix")
```
Is it too late to sync up these names?

wetneb replied:
It is worth making the code and the JSON format consistent but I would advocate to keep the `@JsonProperty` annotations anyway.

tfmorris commented on extensions/wikidata/src/org/openrefine/wikidata/utils/EntityCache.java:
```diff
@@ -39,24 +39,23 @@
-                    public EntityDocument load(String entityId)
-                            throws Exception {
-                        EntityDocument doc = _fetcher.getEntityDocument(entityId);
+                    public EntityDocument load(String entityId) throws Exception {
```
This is a very wide `throws` statement. Can it be narrowed at all to a smaller subset of `Exception`s?

tfmorris commented on extensions/wikidata/tests/data/manifest/wikidata-manifest-v1.0.json:
```diff
@@ -0,0 +1,166 @@
+{
```
As I mentioned by email, if we think users are going to have to deal with manifests, we may want to think about using YAML instead of JSON. In addition to be easier for humans to edit, it allows comments.

tfmorris commented on extensions/wikidata/tests/src/org/openrefine/wikidata/manifests/ManifestTest.java:
```diff
@@ -0,0 +1,34 @@
+
+        Constraints constraints = manifest.getConstraints();
+
+        // TODO: more tests
```
In particular, we need tests for version (missing, different minor version, different major version), as well as malformed manifests.

afkbrb commented on extensions/wikidata/src/org/openrefine/wikidata/commands/LoginCommand.java:
```diff
@@ -79,29 +80,37 @@ public void doPost(HttpServletRequest request, HttpServletResponse response)
 
-        if (isBlank(username) && isBlank(password) && isBlank(consumerToken) &&
-                isBlank(consumerSecret) && isBlank(accessToken) && isBlank(accessSecret)) {
+        if (isBlank(wikibaseApiEndpoint) && isBlank(username) && isBlank(password) && isBlank(consumerToken)
```
A Wikibase instance may don't have [the OAuth extension](https://www.mediawiki.org/wiki/Extension:OAuth) installed, so password based logins should be supported.

wetneb replied:
By the way, MediaWiki makes it possible to create "bot passwords", which are alternative passwords for the account that also come with restricted permissions. We should encourage users to create bot passwords to use them from OpenRefine: https://addshore.com/2020/07/using-openrefine-with-wikidata-for-the-first-time/

afkbrb replied:
To handle "bot passwords", do we need to change the current code?

wetneb replied:
No, it is the same API - we could just consider tweaking the UI to encourage users to create such passwords. Arguably that is a different issue, independent from this PR.

tfmorris replied:
Yes, we should definitely **strongly** encourage this. It looks like "bot" usernames can be identified by the @ sign that they contain, so we can issue strong warning if we don't see it (and also encourage them to use OAuth if available). 

I'm fine with that being a separate PR, as long as we don't lose track of it.

afkbrb commented on extensions/wikidata/src/org/openrefine/wikidata/commands/PreviewWikibaseSchemaCommand.java:
```diff
@@ -81,14 +84,29 @@ public void doPost(HttpServletRequest request, HttpServletResponse response)
+                try {
+                    manifest = ManifestParser.parse(manifestJson);
+                } catch (ManifestException e) {
+                    respondError(response, "Wikibase manifest could not be parsed.");
```
Maybe I should append `e.getMessage()` to the message.

tfmorris replied:
Yes, that would be good. There's also `org.apache.commons.lang.exception.ExceptionUtils.getFullStackTrace(e)`, but hopefully the stacktrace won't be necessary because the range of errors here is limited.

afkbrb commented on extensions/wikidata/src/org/openrefine/wikidata/manifests/ManifestParser.java:
```diff
@@ -0,0 +1,37 @@
+            logger.warn("version is missing in the manifest, assumed to be \"1.0\"");
+        }
+
+        if ("1.0".equals(version)) {
```
I didn't deal with versioning well in this prototype, and I'm not sure if it's necessary to have the manifest versioned. I'll add a comment below to talk about my thinking on the versioning.

afkbrb commented on extensions/wikidata/src/org/openrefine/wikidata/manifests/constraints/ValueTypeConstraint.java:
```diff
@@ -0,0 +1,54 @@
+        this.relation = relation;
+        this.instanceOf = instanceOf;
+        this.subclassOf = subclassOf;
+        this.instanceOrSubclassOf = instanceOrSubclassOf;
```
I'll fix this.

afkbrb commented on extensions/wikidata/src/org/openrefine/wikidata/manifests/v1_0/ConstraintsV1_0.java:
```diff
@@ -0,0 +1,190 @@
+        // underscore style -> camel style
+        mapper.setPropertyNamingStrategy(PropertyNamingStrategy.SNAKE_CASE);
+
+        ArrayNode constraints = (ArrayNode) node;
+        for (Iterator<JsonNode> it = constraints.elements(); it.hasNext(); ) {
+            JsonNode constraint = it.next();
```
fixed

afkbrb commented on extensions/wikidata/src/org/openrefine/wikidata/manifests/v1_0/ConstraintsV1_0.java:
```diff
@@ -0,0 +1,190 @@
+    private Map<String, Constraint> nameToConstraint = new HashMap<>();
+
+    static {
+        nameToConstraintClass.put("allowed_entity_types_constraint", AllowedEntityTypesConstraint.class);
```
I'll give it a try.

afkbrb commented on extensions/wikidata/src/org/openrefine/wikidata/qa/scrutinizers/EditScrutinizer.java:
```diff
@@ -57,6 +54,19 @@ public void setStore(QAWarningStore store) {
+     *
+     * @return false if any necessary dependency is missing, true otherwise.
+     */
+    public abstract boolean prepareDependencies();
```
> I'm not sure I fully understand the lifecycle here, but could this be rolled into `batchIsBeginning()`? 

I don't think so. `batchIsBeginning()`, `scrutinize()` and `batchIsFinished()` are all parts of the scrutinization actually, see [NoEditsMadeScrutinizer](https://github.com/OpenRefine/OpenRefine/blob/master/extensions/wikidata/src/org/openrefine/wikidata/qa/scrutinizers/NoEditsMadeScrutinizer.java). My thought is that if `prepareDependencies` fails, the scrutinizer should not be registered at all, see my code from [EditInspector](https://github.com/afkbrb/OpenRefine/blob/wikibase-support/extensions/wikidata/src/org/openrefine/wikidata/qa/EditInspector.java) below:

```java
    public void register(EditScrutinizer scrutinizer) {
        scrutinizer.setStore(warningStore);
        scrutinizer.setFetcher(fetcher);
        scrutinizer.setConstraints(constraints);
        if (scrutinizer.prepareDependencies()) {
            String key = scrutinizer.getClass().getName();
            scrutinizers.put(key, scrutinizer);
        }
    }
```

> is it something that we want subclasses to explicitly think about and handle?

Yes. 

> Under what circumstances `prepareDependencies` fail? What should the user do about it?

`prepareDependencies ` fails if any dependency needed is missing (and this is expected). In this case, the scrutinizer is skipped since it can't work properly without dependencies. The user is not involved here.

afkbrb commented on extensions/wikidata/src/org/openrefine/wikidata/qa/scrutinizers/ConflictsWithScrutinizer.java:
```diff
@@ -44,6 +44,16 @@
 
+    @Override
+    public boolean prepareDependencies() {
+        if (constraints.getConflictsWithConstraint() == null) return false;
```
Got it!

afkbrb commented on extensions/wikidata/src/org/openrefine/wikidata/qa/scrutinizers/DifferenceWithinRangeScrutinizer.java:
```diff
@@ -23,9 +23,9 @@
-                List<Value> lowerValueProperty = findValues(specs, DIFFERENCE_WITHIN_RANGE_CONSTRAINT_PID);
-                List<Value> minValue = findValues(specs, MINIMUM_VALUE_PID);
-                List<Value> maxValue = findValues(specs, MAXIMUM_VALUE_PID);
+                List<Value> lowerValueProperty = findValues(specs, differenceWithinRangeConstraintPid);
```
Sorry I don't get what you mean here.

afkbrb commented on extensions/wikidata/src/org/openrefine/wikidata/qa/scrutinizers/DifferenceWithinRangeScrutinizer.java:
```diff
@@ -39,6 +39,16 @@
+        differenceWithinRangeConstraintPid = constraints.getDifferenceWithinRangeConstraint().getProperty();
+        minimumValuePid = constraints.getDifferenceWithinRangeConstraint().getMinimumValue();
+        maximumValuePid = constraints.getDifferenceWithinRangeConstraint().getMaximumValue();
+        return differenceWithinRangeConstraintQid != null && differenceWithinRangeConstraintPid != null && minimumValuePid != null && maximumValuePid != null;
```
Got it.

afkbrb commented on extensions/wikidata/src/org/openrefine/wikidata/qa/EditInspector.java:
```diff
@@ -76,15 +80,20 @@ public EditInspector(QAWarningStore warningStore) {
         scrutinizer.setStore(warningStore);
         scrutinizer.setFetcher(fetcher);
+        scrutinizer.setConstraints(constraints);
+        if (scrutinizer.prepareDependencies()) {
```
I think we should log the failures here. Causing an error is no proper since missing dependencies is expected.

afkbrb commented:
Design questions:

1. How to add a new Wikibase manifest? We have two options: 

- Adding via URL. If the MediaWiki API endpoint can exposes configuration of extensions (i.e. https://phabricator.wikimedia.org/T155155 gets finished, but it seems that this can still take a long time), it **may** be possible to add the manifest via the MediaWiki API. I use "may" because that's not enough. For example, the reconciliation service is an external service instead of an extension to a Wikibase instance, so it will not be included in the configuration of the extensions.

- Adding via JSON/YAML text. We can provide manifests for some popular Wikibase instances on our wiki (but we need to run reconciliation services for them first), and write a tutorial to help users to write manifests for their own Wikibase instances. We should encourage users to share the manifests they write for public Wikibase instances on our wiki. To add a manifest, in most cases, the user can just copy and paste the manifest provided on our wiki, and this is simple as copying and pasting a URL. The advantage of this option is that if a user runs a Wikibase locally, he doesn't need to host the manifest as a service and add the URL of it to OpenRefine, he can paste the manifest directly to OpenRefine.

For reference: @wetneb 's comment under https://phabricator.wikimedia.org/T197588

> One other approach to this problem would be to consider that these manifest files are not expected to be necessarily hosted by the Wikibase instance itself - these configuration files could be user-contributed and hosted anywhere (or derived automatically from the Wikibase Registry). The downside is that this requires more work from the community (users need to maintain these manifest files themselves) but it could be necessary if we want to include things like URLs of external tools like QuickStatements.

The latter option is somewhat like the "one other approach" described in @wetneb 's comment, except that the manifest doesn't need to be hosted as a service.

2. Is it necessary to include versioning in the manifest? I think we can just set a standard to the format of the manifest, just like the reconciliation servcie API manifest: https://wdreconcile.toolforge.org/en/api.

wetneb commented:
I would intuitively favour adding a manifest by URL instead of pasting it in a text area, for the following reasons:
- it feels cleaner if users do not have to copy/paste JSON/YAML themselves: they should not have to know about this configuration format to use a third-party Wikibase instance;
- adding a manifest by URL makes it easier to pick up updates: if the target Wikibase adds support for another constraint, changes the URL of its reconciliation service, or anything else, we could re-fetch the manifest from the URL to benefit from these updates. This is something we are not doing yet for reconciliation endpoints, but as a service maintainer I would really like it to be supported (for instance, the absence of refresh makes it impossible for me to change the size of the preview iframe, since its dimensions are stored in the manifest).
- down the line, if we can convince Wikibase itself to serve this data from their API, it would be even easier for users

That being said it does bring up other issues:
- if I want to craft a manifest for a Wikibase instance that has not got one already, I need to somehow host it somewhere if we can only add it in OpenRefine by URL.
- one way to share manifests could be to put them on the target wiki itself, but then we would need to make sure the JSON source can be fetched from there (without any HTML around it) - not sure if there is an easy way to do that in MediaWiki by default.

So perhaps my vote goes towards supporting both methods, as your current UI suggests. The text field could be collapsed by default so that we encourage the use of URLs more? For automatic updates to be possible we would need to record the URL in the backend (not just the manifest contents).

afkbrb commented:
@wetneb I have some questions on dealing with versioning of the manifest format.

Form the roadmap you drafted [here](https://github.com/OpenRefine/OpenRefine/issues/1640#issue-329822161):

> Draft and document an initial version of the manifest format. The format should include versioning so that we can switch to something lighter once MediaWiki exposes more info. 

1. I can't get what you mean here. Do you mean that we can design a manifest that includes anything OpenRefine needs to know about the target Wikibase for now, and once the MediaWiki API exposes more configurations of the extensions, we can delete some information form the manifest and fetch them via the MediaWiki API instead?

2. Do we need to handle manifests with different minor versions with different code? Since minor changes are supposed to be backward compatible, I think one code for one major version is enough.

3. When may backward incompatible changes take place? If that won't happen, we don't need to include versioning in the manifest format at all.

wetneb commented:
> I can't get what you mean here. Do you mean that we can design a manifest that includes anything OpenRefine needs to know about the target Wikibase for now, and once the MediaWiki API exposes more configurations of the extensions, we can delete some information form the manifest and fetch them via the MediaWiki API instead?

Yes exactly!

In general, I expect the manifest format to evolve with the addition of new fields. That alone does not require versioning: we can just make sure our parsing code ignores unknown properties, and provides sensible defaults when newly introduced properties are not supplied.

If one day an API action is introduced to expose the configuration of the constraint system, we could potentially remove the constraint definitions from the manifest. For such a move, I think we would need a new version of the manifest format, such that an old OpenRefine does not accept the new manifests (as using them would disable all QA checks, since the old instance would not be able to fetch constraint definitions).

Another scenario would be the creation of a manifest format which would be used by other tools as well - in that context we could be forced to change terminology, format, and so on (because we are working with other teams with other constraints and conventions). The new manifests would not be parseable directly by existing OpenRefine instances.

So for these use cases I would tend to have OpenRefine refuse to import manifests in newer formats (only looking at the major version, if we are using semver). That's just on top of my mind, I do not claim I have thought about this issue very deeply - I am sure you have a better understanding of the issue!

> Do we need to handle manifests with different minor versions with different code? Since minor changes are supposed to be backward compatible, I think one code for one major version is enough.

I would say so.

tfmorris commented:
> When may backward incompatible changes take place? If that won't happen, we don't need to include versioning in the manifest format at all

Hopefully never, but it's much better to have versioning and not need it, than the opposite. Protocols and file formats should always be versioned, if possible.

tfmorris commented on extensions/wikidata/src/org/openrefine/wikidata/manifests/constraints/AllowedEntityTypesConstraint.java:
```diff
@@ -0,0 +1,67 @@
+    private String wikibaseMediainfo;
+
+    @JsonCreator
+    public AllowedEntityTypesConstraint(@JsonProperty("qid") String qid,
```
OK

afkbrb commented:
After changing [translation-en.json](https://github.com/OpenRefine/OpenRefine/blob/master/extensions/wikidata/module/langs/translation-en.json), do I need to update the translations in other languages?

wetneb commented:
No, outdated translations will be flagged to translators by Weblate directly.

afkbrb commented:
That's great!

tfmorris commented on extensions/wikidata/module/scripts/menu-bar-extension.js:
```diff
@@ -1,134 +1,153 @@
-	  module : "wikidata",
+  url: "command/core/load-language?",
+  type: "POST",
+  async: false,
```
It would be good if this didn't block the UI (but I know it's probably not your code).

tfmorris commented on extensions/wikidata/module/scripts/dialogs/wikibase-dialog.js:
```diff
@@ -0,0 +1,128 @@
+  elmts.addButton.click(function () {
+    let addManifest = function (manifest) {
+      if (!manifest) {
+        alert("manifest is empty");
```
Can these error messages be internationalized?

afkbrb replied:
The work on checking the format of the manifest is not finished yet, I'll add i18n support when this part is done.

tfmorris replied:
Thanks for letting me know that it will be addressed. I'm trying to strike a balance between commenting constantly and ignoring you. :) I'm also concerned that the PR is becoming enormous and will be completely unreviewable if I don't do it incrementally.

tfmorris commented on extensions/wikidata/module/scripts/menu-bar-extension.js:
```diff
@@ -1,134 +1,153 @@
+  $(form).css("display", "none")
+      .attr("method", "post")
+      .attr("action", "command/core/export-rows/" + targetUrl)
+      .attr("target", "gridworks-export-" + format);
```
"Gridworks" is the name that preceded "Google Refine", which was replaced by "OpenRefine" so occurrences of "gridworks" should be globally replaced with "openrefine"

tfmorris commented on extensions/wikidata/module/scripts/wikibase-manager.js:
```diff
@@ -0,0 +1,168 @@
+        if (toUpdate) {
+          // wait for 10s for fetching latest manifests to finish
+          setTimeout(function () {
+            WikibaseManager.saveWikibases();
```
Can this be put in the Ajax completion function rather than relying on a fixed timer?

tfmorris commented on extensions/wikidata/src/org/openrefine/wikidata/manifests/ManifestParser.java:
```diff
@@ -4,7 +4,7 @@
 import com.fasterxml.jackson.databind.ObjectMapper;
 import org.apache.commons.lang3.StringUtils;
-import org.openrefine.wikidata.manifests.v1_0.ManifestV1_0;
+import org.openrefine.wikidata.manifests.v1.ManifestV1;
```
Probably don't need the extra level of package depth here (ie remove ".v1")

afkbrb commented:
Emm, this is incorrect, I'll fix it tomorrow :(

afkbrb commented:
@tfmorris @wetneb I'm planning to write a JSON schema and use [ajv](https://github.com/ajv-validator/ajv) (MIT license) to validate the format of the manifest. To use ajv, can I just copy [ajv.min.js](https://cdnjs.cloudflare.com/ajax/libs/ajv/6.12.3/ajv.min.js) to our src code?

wetneb commented:
That is unfortunately how we "manage" frontend dependencies indeed!

About ajv itself, I would watch out for the time it takes to compile the schema. My understanding is that this library is designed to make validation very quick, potentially at the expense of a higher initialization cost. For instance, if compilation is done in the main thread, it could potentially slow down the UI. Perhaps there is a way to compile a schema only once and then store the compiled validator as a JS file? (That could well be over-engineered, too) Or just compile the schema lazily, only when a validation is first required.

afkbrb commented:
> Perhaps there is a way to compile a schema only once and then store the compiled validator as a JS file?

Yes, see https://github.com/ajv-validator/ajv#getting-started.

Also, the validation only happens when a manifest is added, which is not frequent at all! So I don't think the time taken to compile the schema would be an issue here.

wetneb commented:
> Also, the validation only happens when a manifest is added, which is not frequent at all! So I don't think the time taken to compile the schema would be an issue here.

Exactly - I just wanted to make sure we do not compile the schema every time the Wikidata extension is loaded.

afkbrb commented:
> > Perhaps there is a way to compile a schema only once and then store the compiled validator as a JS file?
> 
> Yes, see https://github.com/ajv-validator/ajv#getting-started.

I just realized that I misunderstood your question. I didn't find a way to store the compiled validator as a JS file. https://github.com/ajv-validator/ajv#getting-started is about compiling a schema to a validator function, so we can reuse the function later when new manifest are added. 

> Exactly - I just wanted to make sure we do not compile the schema every time the Wikidata extension is loaded.

I'm afraid the compiling is necessary, I'll check the time needed.

wetneb commented:
> I'm afraid the compiling is necessary, I'll check the time needed.

As long as it can be deferred to, say, the initialization of the dialog where a manifest can be added, it should not be a problem.

afkbrb commented:
Yes, it can be deferred!

afkbrb commented:
@wetneb Currently in [the manifest of Wikidata](https://github.com/afkbrb/OpenRefine/blob/wikibase-support/extensions/wikidata/tests/data/manifest/wikidata-manifest-v1.0.json#L83), the reconciliation service endpoint is "https://wdreconcile.toolforge.org/en/api", which is hard-coded for a specific language ("en").  This is the same in the [gist](https://gist.github.com/despens/d6ae4110c4e97944ddba29f23d78899f#file-manifest-js-L61). I think it's better to include a "${lang}" variable in the URL (change the URL to "https://wdreconcile.toolforge.org/${lang}/api"), so we can support other languages. What's your opinion on this?

wetneb commented:
@afkbrb that would make a lot of sense, thanks for taking care of this!

afkbrb commented:
I've finished supporting fetching different language codes **of monolingual texts** for different Wikibases in the backend with [this commit](https://github.com/OpenRefine/OpenRefine/pull/2810/commits/836ee53e93a7caf8530661941c209fd756e48d74#diff-8c945dbeb162674cfbbd7ad7010ea770). But I just found that I should do that in [langsuggest.js](https://github.com/OpenRefine/OpenRefine/blob/master/extensions/wikidata/module/scripts/langsuggest.js), too. I can do it today.

@wetneb In both [LanguageCodeStore](https://github.com/OpenRefine/OpenRefine/blob/master/extensions/wikidata/src/org/openrefine/wikidata/utils/LanguageCodeStore.java#L13) and [langsuggest.js](https://github.com/OpenRefine/OpenRefine/blob/master/extensions/wikidata/module/scripts/langsuggest.js#L15), you mentioned that "in the future, store different lists for term languages and monolingual text languages." Should this also be done in this PR?

wetneb commented on extensions/wikidata/src/org/openrefine/wikidata/commands/PreviewWikibaseSchemaCommand.java:
```diff
@@ -105,10 +105,12 @@ public void doPost(HttpServletRequest request, HttpServletResponse response)
-            EditInspector inspector = new EditInspector(warningStore, manifest);
-            inspector.inspect(editBatch);
-            
+            if (manifest.getConstraintsRelatedId("property_constraint_pid") != null) {
```
Not all scrutinizers rely on WikibaseQualityConstraints, so it is worth inspecting the batch even if no constraints are set up in the manifest.

afkbrb replied:
Fixed with commit 4d9cc4e below.

afkbrb commented:
@wetneb 's comment under #2304:
> As part of #1640 we want to make the configuration of the Wikibase instance user-configurable (instead of being hard-coded to Wikidata) and part of that includes letting users set the maxlag value as they wish.

@wetneb Did you mean that we should include the default maxlag in the manifest? So the actual maxlag used will be the maxlag  in the manifest unless the key "wikibase.upload.maxLag" is set in the preferences?  Should we use different keys for different Wikibases (e.g. change the value of "wikibase.upload.maxLag" to a map), so that the user can set maxlag for each Wikibase respectively?

wetneb commented:
@afkbrb good question! I am not too sure. Perhaps the right place to configure this is not the global preferences, but rather when running an upload operation (this could be added alongside the summary, for instance).
In the meantime, I think your first option makes sense:

>  So the actual maxlag used will be the maxlag in the manifest unless the key "wikibase.upload.maxLag" is set in the preferences?

So manifests would just advertise the default maxlag for the instance.

afkbrb commented:
> Perhaps the right place to configure this is not the global preferences, but rather when running an upload operation (this could be added alongside the summary, for instance).

Good idea! So we can add the default maxlag in the manifest. The user can choose to override it when uploading edits. It also helps avoid the complexity of setting maxlag of each Wikibase in the preferences.

thadguidry commented on extensions/wikidata/module/scripts/menu-bar-extension.js:
```diff
@@ -1,134 +1,153 @@
+        "submenu": [
+          {
+            id: "wikidata/manager-wikibase",
+            label: "Manage Wikibase",
```
Configure Wikibase account?

wetneb replied:
Could be better!

tfmorris commented:
> could you add to your workflow the update of the release notes

Sure, I can do that. I'm used to writing the release notes at the end, but doing them in a running fashion works for me too.
